### DISCUSSION TITLE
Can't get openssl to change openssl_conf directory

### DISCUSSION THREAD

**Discussion by ilburdma-cisco** [openssl/openssl#25602] [Q&A]
Hello all,
I am trying to build openssl with fips support in a docker container using latest alpine and openssl images.
openssl builds fine with proper FIPS support however i do not see it in the providers.

This is the Configure command i run prior to build
```
./Configure enable-fips --prefix=/usr/local/ssl --openssldir=/usr/local/ssl
```
Successful fips build
```
/usr/local/ssl # bin/openssl fipsinstall -out /usr/local/ssl/fipsmodule.cnf -module /usr/local/ssl/lib/oss
l-modules/fips.so -pedantic
HMAC : (KAT_Integrity) : Pass
HMAC : (Module_Integrity) : Pass
SHA1 : (KAT_Digest) : Pass
SHA2 : (KAT_Digest) : Pass
SHA3 : (KAT_Digest) : Pass
AES_GCM : (KAT_Cipher) : Pass
AES_ECB_Decrypt : (KAT_Cipher) : Pass
RSA : (KAT_Signature) : Pass
ECDSA : (KAT_Signature) : Pass
ECDSA : (KAT_Signature) : Pass
DSA : (KAT_Signature) : Pass
TLS13_KDF_EXTRACT : (KAT_KDF) : Pass
TLS13_KDF_EXPAND : (KAT_KDF) : Pass
TLS12_PRF : (KAT_KDF) : Pass
PBKDF2 : (KAT_KDF) : Pass
SSHKDF : (KAT_KDF) : Pass
KBKDF : (KAT_KDF) : Pass
KBKDF_KMAC : (KAT_KDF) : Pass
HKDF : (KAT_KDF) : Pass
SSKDF : (KAT_KDF) : Pass
X963KDF : (KAT_KDF) : Pass
X942KDF : (KAT_KDF) : Pass
HASH : (DRBG) : Pass
CTR : (DRBG) : Pass
HMAC : (DRBG) : Pass
DH : (KAT_KA) : Pass
ECDH : (KAT_KA) : Pass
RSA_Encrypt : (KAT_AsymmetricCipher) : Pass
RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
        name:           OpenSSL FIPS Provider
        version:        3.3.2
        build:          3.3.2
INSTALL PASSED
```
when I try to list providers FIPS provider is not there.

```
/usr/local/ssl # openssl list -providers
Providers:
  default
    name: OpenSSL Default Provider
    version: 3.3.2
    status: active
```
When I run version check I can see OPENSSLDIR is set to ```/etc/ssl/``` instead of ```/usr/local/ssl/bin```
```
/usr/local/ssl # bin/openssl version -a
OpenSSL 3.3.2 3 Sep 2024 (Library: OpenSSL 3.3.2 3 Sep 2024)
built on: Tue Sep  3 18:09:57 2024 UTC
platform: linux-aarch64
options:  bn(64,64)
compiler: gcc -fPIC -pthread -Wa,--noexecstack -Os -fstack-clash-protection -Wformat -Werror=format-security -g -Os -fstack-clash-protection -Wformat -Werror=format-security -g -Wa,--noexecstack -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
OPENSSLDIR: "/etc/ssl"
ENGINESDIR: "/usr/lib/engines-3"
MODULESDIR: "/usr/lib/ossl-modules"
Seeding source: os-specific
CPUINFO: OPENSSL_armcap=0x8fd
/usr/local/ssl # 
```
I tried to change OPENSSL_CONF environment variable but openssl is not picking it up. 
```
/usr/local/ssl # echo $OPENSSLDIR
/usr/local/ssl
```

I am not sure what do to next

Any help will be welcomed.

**Comment by nhorman** [comment]
I _think_ (having trouble remembering), but I believe that openssldir is rooted in prefix, ie your configuration:
```
./Configure enable-fips --prefix=/usr/local/ssl --openssldir=/usr/local/ssl
```
at install time will install libs/binaries/etc to /usr/local/ssl

but will place openssldir in /usr/local/ssl/usr/local/ssl

What it sounds like you want is:
```
/Configure enable-fips --prefix=/usr/local/ssl --openssldir=/
```

**Reply by t8m** [reply]
Actually no, only if the path passed to `--openssldir` is relative, the prefix is prepended. Otherwise the value is used as is.

**Comment by t8m** [comment]
> When I run version check I can see OPENSSLDIR is set to `/etc/ssl/` instead of `/usr/local/ssl/bin`
> 
> ```
> /usr/local/ssl # bin/openssl version -a
> OpenSSL 3.3.2 3 Sep 2024 (Library: OpenSSL 3.3.2 3 Sep 2024)
> built on: Tue Sep  3 18:09:57 2024 UTC
> platform: linux-aarch64
> options:  bn(64,64)
> compiler: gcc -fPIC -pthread -Wa,--noexecstack -Os -fstack-clash-protection -Wformat -Werror=format-security -g -Os -fstack-clash-protection -Wformat -Werror=format-security -g -Wa,--noexecstack -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
> OPENSSLDIR: "/etc/ssl"
> ENGINESDIR: "/usr/lib/engines-3"
> MODULESDIR: "/usr/lib/ossl-modules"
> Seeding source: os-specific
> CPUINFO: OPENSSL_armcap=0x8fd
> /usr/local/ssl # 
> ```

This sounds like you have 2 instances of openssl (3.3.2) installed in the system. One in /usr/local/ssl and one in /usr/lib. The invocation of /usr/local/ssl/bin/openssl still links to the libcrypto/libssl from /usr/lib. Try setting `export LD_LIBRARY_PATH=/usr/local/ssl/lib`.

**Comment by ilburdma-cisco** [comment]
Unfortunately, I am still failing.
I exported the environment variable, added the directory to the path, and renewed the session. Still getting the same default provider.

I forgot to mention that i am using Alpine for this build

**Comment by Sashan** [comment]
How is your PATH variable set? is your shell picking up the right openssl binary? What does `which openssl` says when running in shell?

**Comment by ilburdma-cisco** [comment]
I set the env variable during the docker build. 

Here is the Dockerfile.
```
# Use Alpine as the base image
FROM alpine:latest

RUN apk add --no-cache \
    build-base \
    wget \
    perl \
    cmake \
    musl-dev \
    linux-headers

RUN wget -q https://github.com/openssl/openssl/releases/download/openssl-3.3.2/openssl-3.3.2.tar.gz  && \
    tar -xzvf openssl-3.3.2.tar.gz && \
    cd openssl-3.3.2 && \
    sed -i 's|#include <linux/mman.h>|#include <sys/mman.h>|' crypto/mem_sec.c && \
    ./Configure enable-fips --prefix=/usr/local/ssl --openssldir=/usr/local/ssl && \
    make -s -j$(nproc) && make -s install_sw

RUN /usr/local/ssl/bin/openssl fipsinstall -out /openssl-3.3.2/providers/fipsmodule.cnf \
    -module /openssl-3.3.2/providers/fips.so -provider_name fips 

ENV PATH="/usr/local/ssl/bin:${PATH}"
ENV LD_LIBRARY_PATH="/usr/local/ssl/lib:${LD_LIBRARY_PATH}"

RUN /usr/local/ssl/bin/openssl version -a && \
    /usr/local/ssl/bin/openssl list -providers -verbose
```

I also attached build.log. I can see in the build log that the ENV variables are set correctly:
```
OPENSSLDIR: "/usr/local/ssl"
ENGINESDIR: "/usr/local/ssl/lib/engines-3"
MODULESDIR: "/usr/local/ssl/lib/ossl-modules"
```
but openssl still refuses to see or use the new providers.

The ```sed``` line is needed to switch include to musl library instead of glibc. 

[build.log](https://github.com/user-attachments/files/17294676/build.log)

**Comment by Sashan** [comment]
I think you also need to update openssl.cnf. Note there lines
```
45 # For FIPS
46 # Optionally include a file that is generated by the OpenSSL fipsinstall
47 # application. This file contains configuration data required by the OpenSSL
48 # fips provider. It contains a named section e.g. [fips_sect] which is
49 # referenced from the [provider_sect] below.
50 # Refer to the OpenSSL security policy for more information.
51 # .include fipsmodule.cnf
```
you need to un-comment line 51 and there is also `[provider_sect]` which needs your attention too. Note you may need to specify absolute path to your `fipsmodule.cnf`. in your case it should be `/openssl-3.3.2/providers/fipsmodule.cnf`
```
 56 # List of providers to load
 57 [provider_sect]
 58 default = default_sect
 59 # The fips section name should match the section name inside the
 60 # included fipsmodule.cnf.
 61 # fips = fips_sect
```
you also need to un-comment line 61.

hope it helps
